### PR TITLE
Allow pooled embeddings on any model

### PR DESCRIPTION
**Pull request by iamlemec**
This allows one to compute pooled embeddings on any model, not just classical embedding models. This is increasingly useful due to the rise of generative-type models in embedding benchmarks (most recently, `gte-Qwen1.5-7B-instruct`). The main changes are:

- Add an `append_pooling` function to `llm_build_context` that grafts a pooling layer onto the last tensor of an existing graph. This makes some assumptions about how the underlying graph is laid out, but we're already doing that in a couple of places, and there are tensor name checks too.
- Add a `LLAMA_POOLING_TYPE_LAST` pooling type since this is a common type of pooling used with generative models. Works very similarly to CLS pooling.
- Allow user to specify attention type on context creation (causal, non-causal, or default for model).
-  Updates `embedding`/`retreival` examples to request correct logits depending on `pooling_type`.

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -7041,6 +7041,44 @@ struct llm_build_context {
         return lctx.inp_s_seq;
     }
 
+    struct ggml_cgraph * append_pooling(struct ggml_cgraph * gf) {
+        struct ggml_tensor * inp = gf->nodes[gf->n_nodes - 1];
+        if (strcmp(inp->name, "result_embd") != 0) {
+            inp = gf->nodes[gf->n_nodes - 2];
+            GGML_ASSERT(strcmp(inp->name, "result_norm") == 0 && "embeddings tensor not found");
```
Comment (compilade) [inline llama.cpp:7048]: This probably won't work for Grok, Phi 2, MiniCPM, and Command R, as their "result_norm" is the 3rd (or sometimes 4th for Command R) last tensor.

File: llama.cpp [hunk 2]
```diff
@@ -11322,30 +11372,13 @@ static int llama_decode_internal(
...
-                gf->n_nodes = i_embd + 1;
+            res = nullptr; // do not extract logits for embedding case
+            embd = gf->nodes[gf->n_nodes - 1];
+            if (strcmp(embd->name, "result_embd_pooled") != 0) {
+                embd = gf->nodes[gf->n_nodes - 2];
             }
+            GGML_ASSERT(strcmp(embd->name, "result_embd_pooled") == 0 && "missing embeddings tensor");
         } else {
             embd = nullptr; // do not extract embeddings when not needed
             GGML_ASSERT(strcmp(res->name, "result_output") == 0 && "missing result_output tensor");
```
Comment (compilade) [inline llama.cpp:11384]: So an embeddings model will crash on the first decode when `cparams.embeddings` is set to `false`?

File: examples/embedding/embedding.cpp [hunk 3]
```diff
@@ -40,13 +56,7 @@ static void batch_decode(llama_context * ctx, llama_batch & batch, float * outpu
 
         // try to get sequence embeddings - supported only when pooling_type is not NONE
         const float * embd = llama_get_embeddings_seq(ctx, batch.seq_id[i][0]);
-        if (embd == NULL) {
-            embd = llama_get_embeddings_ith(ctx, i);
-            if (embd == NULL) {
-                fprintf(stderr, "%s: failed to get embeddings for token %d\n", __func__, i);
-                continue;
-            }
-        }
```
Comment (compilade) [inline examples/embedding/embedding.cpp:49]: Why remove support for `LLAMA_POOLING_TYPE_NONE` in the `embedding` example?

File: examples/retrieval/retrieval.cpp [hunk 4]
```diff
@@ -133,9 +133,25 @@ static std::vector<chunk> chunk_file(const std::string & filename, int chunk_siz
...
+            return pos == 0;
+        case LLAMA_POOLING_TYPE_LAST:
+            return pos == n_tokens - 1;
+        default:
+            GGML_ASSERT(false && "unsupported pooling type");
+    }
+}
+
+static void batch_add_seq(llama_batch & batch, const std::vector<int32_t> & tokens, int seq_id, enum llama_pooling_type pooling_type) {
+    int n_tokens = tokens.size();
```
Comment (compilade) [inline examples/retrieval/retrieval.cpp:151]:
```suggestion
static void batch_add_seq(llama_batch & batch, const std::vector<int32_t> & tokens, llama_seq_id seq_id, enum llama_pooling_type pooling_type) {
    size_t n_tokens = tokens.size();
```

File: examples/embedding/embedding.cpp [hunk 5]
```diff
@@ -17,9 +17,25 @@ static std::vector<std::string> split_lines(const std::string & s) {
...
+            return pos == 0;
+        case LLAMA_POOLING_TYPE_LAST:
+            return pos == n_tokens - 1;
+        default:
+            GGML_ASSERT(false && "unsupported pooling type");
+    }
+}
+
+static void batch_add_seq(llama_batch & batch, const std::vector<int32_t> & tokens, int seq_id, enum llama_pooling_type pooling_type) {
+    int n_tokens = tokens.size();
```
Comment (compilade) [inline examples/embedding/embedding.cpp:35]:
```suggestion
static void batch_add_seq(llama_batch & batch, const std::vector<int32_t> & tokens, llama_seq_id seq_id, enum llama_pooling_type pooling_type) {
    size_t n_tokens = tokens.size();
```

File: llama.cpp (hunk 2 above)
Comment (compilade) [inline llama.cpp:11384]: There are places that need to know when embeddings or logits will be output, like `llama_output_reserve`

https://github.com/ggerganov/llama.cpp/blob/cd93a28cb1446319af5e2f4b416174c3a8e43546/llama.cpp#L11064-L11065

This will need to be updated to reflect exactly how this affects what happens later in this function near the comments `// extract logits` and `// extract embeddings`.

**Review by compilade — COMMENTED** [review]
Interesting and seems like this is going to be useful. Otherwise only BERT was affected by pooling types.

File: llama.cpp (hunk 1 above)
Comment (iamlemec) [reply llama.cpp:7048]: Indeed, brining back the backwards search for `result_norm`.

File: llama.cpp (hunk 2 above)
Comment (iamlemec) [reply llama.cpp:11384]: Yeah, though I can't think of any case where you'd use an embedding model without `cparams.embeddings`. I guess there's nothing really indicating something is an embedding model other than the lack of a `result_output` tensor, so it's hard to intercept this earlier and give an error.

File: examples/embedding/embedding.cpp (hunk 3 above)
Comment (iamlemec) [reply examples/embedding/embedding.cpp:49]: Mostly because we're not actually printing out the entire token level embeddings anyway. The way it was implemented before was essentially doing last token pooling (not necessarily the last position in the sequence though, just the last one in the order the batch was loaded), but now that last token pooling is an official option, may as well encourage the user to make that choice conciously.

File: llama.cpp (hunk 2 above)
Comment (iamlemec) [reply llama.cpp:11384]: So can we get away with saying you're either getting logits or embeddings but never both, and that behavior is exclusively controlled by `cparams.embeddings`? In that case we could just have
```c++
const bool has_logits = !cparams.embeddings;
const bool has_embd   =  cparams.embeddings;
```

Comment (compilade) [reply llama.cpp:11390]: Well, `hparams.causal_attn` is `false` for BERT at least, and it's the only embedding-only model architecture currently in `llama.cpp`. All BERT-like architectures also set this key to `false` when converted to GGUF.

There might be a need for a dedicated metadata key-value pair for embedding-only models if non-causal text generation models are a thing. (T5? Or is it causal?) Anyway, `cparams.causal_attn` can be used to get non-causal attention with any model, I think (I did not test this), except for recurrent models (Mamba).

I think there should at least be some abstraction (exported in `llama.h`) to know whether or not a model **can** provide embeddings and/or logits. This would make things like #7448 easier, even if it initially relies on `hparams.causal_attn`.

Comment (compilade) [reply llama.cpp:11390]: Yeah, I can't really think of a use-case where both would be needed at the same time. ***Except*** *maybe* for a `server` serving both completions and embeddings out of the same model. So that's something to consider.

Comment (iamlemec) [reply llama.cpp:11390]: Agreed, but for a given call to `llama_decode` presumably you would never want both. For the `gritlm` example, I actually just made two contexts, one for generation one for embeddings. Another option would be to add a `llama_set_embeddings` function.

Comment (iamlemec) [reply llama.cpp:11390]: I see, so at least for now, it looks like `hparams.causal_attn` is a good indicator of whether a model is embedding-only. And I can't imagine a generative model with non-causal attention. I think T5 is causal, at least for the decoder part.

Then I guess we want to assert `hparams.causal_attn || cparams.embeddings` at some point. That way we don't have to worry about divergence and the error is caught earlier.

File: llama.cpp [hunk 6]
```diff
@@ -11061,8 +11117,8 @@ static size_t llama_output_reserve(llama_context & lctx, size_t n_outputs) {
     const auto n_embd  = hparams.n_embd;
 
     // TODO: use a per-batch flag for logits presence instead
-    const bool has_logits = cparams.causal_attn;
-    const bool has_embd   = cparams.embeddings && (hparams.causal_attn || cparams.pooling_type == LLAMA_POOLING_TYPE_NONE);
+    const bool has_logits = !cparams.embeddings;
+    const bool has_embd   =  cparams.embeddings;
```
Comment (compilade) [inline llama.cpp:11121]: Note that `lctx.embd` is not used by all pooled embeddings types, it's really only used with `LLAMA_POOLING_TYPE_NONE`.

(This is all done near the end of `llama_decode_internal` in a `switch` statement)

So maybe the condition for `has_embd` could be `cparams.embeddings && cparams.pooling_type == LLAMA_POOLING_TYPE_NONE`?

(See also the other places where `hparams.causal_attn` was used to understand the assumptions that stem from it, to check if they need to be modified)

Comment (iamlemec) [reply llama.cpp]: Ah, that makes sense. I put in the `pooling_type` check you suggested in there. I also changed the `inp_out_ids` calculation to rely on `cparams.embeddings` rather than `hparams.causal_attn`.

File: llama.h [hunk 7]
```diff
@@ -275,7 +282,7 @@ extern "C" {
 
         enum llama_rope_scaling_type rope_scaling_type; // RoPE scaling type, from `enum llama_rope_scaling_type`
         enum llama_pooling_type      pooling_type;      // whether to pool (sum) embedding results by sequence id
-                                                        // (ignored if no pooling layer)
+        enum llama_attention_type    attention_type;    // causal, non-causal, or unspecified
```
[the remark this answers is not in the document]
Comment (iamlemec) [reply llama.h:285]: Yeah, actually I don't think it gives you any new capabilities. Perhaps it's best to keep it the way it is and avoid breaking changes. Will switch back!

[the remark this answers is not in the document]
Comment (iamlemec) [reply llama.h:285]: The only inconvenience is that it makes it slightly awkward to specify as an CLI flag in the examples.

[the remark this answers is not in the document]
Comment (iamlemec) [reply llama.h]: Ok, took out the `attention_type` stuff, should work now. If server tests are failing, does that mean I have to rebase on master?

File: llama.h [hunk 8]
```diff
@@ -759,6 +759,10 @@ extern "C" {
     // n_threads_batch is the number of threads used for prompt and batch processing (multiple tokens)
     LLAMA_API void llama_set_n_threads(struct llama_context * ctx, uint32_t n_threads, uint32_t n_threads_batch);
 
+    // Set whether the model is in embeddings model or not
```
Comment (ggerganov) [inline llama.h:762]:
```suggestion
    // Set whether the model is in embeddings mode or not
```

**Review by ggerganov — APPROVED** [review]

**Comment by ngxson** [comment]
@iamlemec This PR looks good and would be very useful. Could you update it to latest `master` to see if it compiles on CI?

Regarding model crash on `warming up the model with an empty run` (see this [CI run](https://github.com/ggerganov/llama.cpp/actions/runs/9416536334/job/25939846528?pr=7477)), do you think it's related to one of the change in this PR?

**Comment by iamlemec** [comment]
@ngxson Just rebased onto master. Let's see if that server error persists. I'm on a weak laptop and internet connection right now, but will double check things later today.

File: llama.cpp [hunk 9]
```diff
@@ -11754,7 +11779,7 @@ static void llama_set_inputs(llama_context & lctx, const llama_batch & batch) {
         ggml_backend_tensor_set(lctx.inp_pos, batch.pos, 0, n_tokens*ggml_element_size(lctx.inp_pos));
     }
 
-    if (hparams.causal_attn || cparams.pooling_type == LLAMA_POOLING_TYPE_NONE) {
+    if (!cparams.embeddings || cparams.pooling_type == LLAMA_POOLING_TYPE_NONE) {
```
Comment (compilade) [inline llama.cpp:11782]: Hmm. Some outputs should still be skipped when embedding for some of the pooling types, no?

This will cause uninitialized `lctx.inp_out_ids` when embedding with non-Bert models with pooling types other than NONE.

This condition was there originally for how BERT managed input skipping.

https://github.com/ggerganov/llama.cpp/blob/f8ec8877b75774fc6c47559d529dac423877bcad/llama.cpp#L8534

Since `batch.logits` is likely correctly set when using pooled embeddings (at least, how you wrote them seems correct), then should this condition instead always be true?

And if that is done, then `inp_cls` would be redundant, since the correct rows would already be the only thing left.

Feel free to make objections.

Comment (iamlemec) [reply llama.cpp:11782]: Yeah, that makes sense. I'm guessing we want to avoid putting a `pooling_type == LLAMA_POOLING_TYPE_NONE` in every single other model? In that case, I guess we have to actually require all logits be set when getting non-NONE embeddings from non-Bert models. The downside is that it results in a needless `get_rows` on all the outputs.

In fact, it seems like `batch.logits` isn't really used when `pooling_type` is not `NONE`, since we use all the outputs and the results are stored in `embd_seq_out`. Or actually, all that's currently required is that at least one logit is requested so you go down the right branch when we check if `lctx.n_outputs == 0` in `llama_decode_internal`. It seems like in this case we might want to officially ignore `batch.logits` and give priority to `cparams.embeddings`.

Comment (compilade) [reply llama.cpp:11782]: I think the simpler way to fix this in the meantime is to make `n_outputs == n_tokens_all` in `llama_decode_internal` for all non-NONE pooling types when `cparams.embeddings` is true. It could then re-use the same logic as `logits_all`.

But I think the CLS and LAST pooling types could eventually skip computing the embeddings they don't need.

Comment (iamlemec) [reply llama.cpp]: Ok, I think this should do it. Basically bypass logits when doing non-NONE embeddings. Note that I'm using `hparams.causal_attn` to decide if we're in a BERT model or not in `llama_set_inputs`.

**Comment by iamlemec** [comment]
@compilade sorry to ping you again, but I think this is ready to go. I believe the issues with `n_outputs` are sorted, and I've tested with a few different models, both embedding and non-embedding.

**Review by compilade — APPROVED** [review]
> sorry to ping you again

Pinging is the right thing, because otherwise I tend to forget to go back and re-review, unless recent activity catches my attention enough to have another look at the changes.

> I believe the issues with `n_outputs` are sorted

I believe this as well.

File: examples/gritlm/gritlm.cpp [hunk 10]
```diff
@@ -44,6 +44,7 @@ static std::vector<std::vector<float>> encode(llama_context * ctx, const std::ve
 
         // clear previous kv_cache values (irrelevant for embeddings)
         llama_kv_cache_clear(ctx);
+        llama_set_embeddings(ctx, true);
```
Comment (ngxson) [inline examples/gritlm/gritlm.cpp:47]: I have a small question here: in the case when both `embeddings` and `causal_attn` are both enabled, will it still be correct?

Comment (iamlemec) [reply examples/gritlm/gritlm.cpp:47]: In general, it's possible to run with `embeddings=true` and `causal_attn=true`, as long as the underlying model supports causal attention. For the GritLM case, I just checked here, and it will run but give incorrect results since it expects the embeddings to be run non-causally.
